### PR TITLE
ESP32-C3: Basic MCUboot support

### PR DESCRIPTION
**Pull request by almir-okato**
This PR adds basic support for ESP32-C3 SoC from Espressif Systems.

**Notes:** 
- Both Zephyr and NuttX RTOSes for ESP32-C3 would be able to boot in the future, however the changes in their respective repositories still are WIP and aren't upstream yet.

### CODE REVIEW

File: boot/espressif/hal/CMakeLists.txt [hunk 1]
```diff
@@ -29,9 +31,16 @@ list(APPEND INCLUDE_DIRS
     ${esp_idf_dir}/components/efuse/include
     ${esp_idf_dir}/components/efuse/${MCUBOOT_TARGET}/include
     )
+
+if("${MCUBOOT_ARCH}" STREQUAL "xtensa")
+    list(APPEND INCLUDE_DIRS
```
Comment (d3zd3z) [inline boot/espressif/hal/CMakeLists.txt:36]: There is stray whitespace at the end of this line.

File: boot/espressif/hal/CMakeLists.txt [hunk 2]
```diff
@@ -92,15 +99,34 @@ set(LDFLAGS
...
 
 set_source_files_properties(
     ${esp_idf_dir}/components/bootloader_support/src/bootloader_flash.c
     ${esp_idf_dir}/components/esp_hw_support/port/${MCUBOOT_TARGET}/rtc_clk_init.c
+    ${esp_idf_dir}/components/esp_hw_support/port/${MCUBOOT_TARGET}/rtc_init.c
     ${esp_idf_dir}/components/esp_hw_support/port/${MCUBOOT_TARGET}/rtc_clk.c
     ${esp_idf_dir}/components/esp_hw_support/port/${MCUBOOT_TARGET}/rtc_time.c
     PROPERTIES COMPILE_FLAGS
     "-Wno-unused-variable")
+
```
Comment (d3zd3z) [inline boot/espressif/hal/CMakeLists.txt:119]: Whitespace also here. Please check for this.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
I think it'd help readability to have spaces after the colons in the subject lines, e.g. (espressif: esp32c3: Add esp32c3 initial basic support).

I'm not really sure how to write it differently, as it is, the two commit texts only differ by a few characters "s2 vs c3". Maybe those used to the target will more easily see the difference. I can't really think of a way to do it differently.

**Comment by gustavonihei** [comment]
> I think it'd help readability to have spaces after the colons in the subject lines, e.g. (espressif: esp32c3: Add esp32c3 initial basic support).
> 
> I'm not really sure how to write it differently, as it is, the two commit texts only differ by a few characters "s2 vs c3". Maybe those used to the target will more easily see the difference. I can't really think of a way to do it differently.

Actually the commit message could be shortened to just:
- `espressif: Add ESP32-S2 initial basic support`
- `espressif: Add ESP32-C3 initial basic support`

The `espressif` tag is meaningful enough for specifying the port name.

**Review by gustavonihei — APPROVED** [review]
LGTM. Also successfully booted a NuttX application firmware image on the ESP32-C3.

File: boot/espressif/hal/CMakeLists.txt (hunk 1 above)
Comment (almir-okato) [reply boot/espressif/hal/CMakeLists.txt:36]: Done

File: boot/espressif/hal/CMakeLists.txt (hunk 2 above)
Comment (almir-okato) [reply boot/espressif/hal/CMakeLists.txt]: Done

**Comment by almir-okato** [comment]
@d3zd3z I've made the changes after your comments, please take a look.

@utzig could you please review this PR as well? It is very similar as the other for ESP32-S2 merged earlier.

File: boot/espressif/hal/CMakeLists.txt (hunk 1 above)
Comment (utzig) [reply boot/espressif/hal/CMakeLists.txt:36]: I still see trailing whitespaces here.

File: boot/espressif/hal/src/esp32c3/bootloader_init.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    uint8_t wp_pin = esp_rom_efuse_get_flash_wp_gpio();
+    uint8_t clk_gpio_num = SPI_CLK_GPIO_NUM;
+    uint8_t q_gpio_num   = SPI_Q_GPIO_NUM;
+    uint8_t d_gpio_num   = SPI_D_GPIO_NUM;
+    uint8_t cs0_gpio_num = SPI_CS0_GPIO_NUM;
+    uint8_t hd_gpio_num  = SPI_HD_GPIO_NUM;
+    uint8_t wp_gpio_num  = SPI_WP_GPIO_NUM;
+    if (spiconfig == 0) {
+
+    } else {
```
Comment (utzig) [inline boot/espressif/hal/src/esp32c3/bootloader_init.c:62]: Test for `spiconfig != 0` and remove the `else`.

File: boot/espressif/port/esp32c3/ld/bootloader.ld [hunk 4]
```diff
@@ -0,0 +1,141 @@
+/*
+ * Linker file used to link the bootloader.
+ */
+ 
+MEMORY
+{
+  iram_seg (RWX) :    org = 0x403CA000, len = 0x6000
+  iram_loader_seg (RWX) : org = 0x403D0000, len = 0x2000
+  dram_seg (RW) :     org = 0x3FCD6000, len = 0x6000
```
Comment (utzig) [inline boot/espressif/port/esp32c3/ld/bootloader.ld:9]: Trailing space here.

**Review by utzig — CHANGES_REQUESTED** [review]
Looks very similar to the previous PR, a few knits, but mostly OK.

File: boot/espressif/hal/CMakeLists.txt (hunk 1 above)
Comment (almir-okato) [reply boot/espressif/hal/CMakeLists.txt]: Sorry, fixed now.

File: boot/espressif/port/esp32c3/ld/bootloader.ld (hunk 4 above)
Comment (almir-okato) [reply boot/espressif/port/esp32c3/ld/bootloader.ld]: Fixed.

File: boot/espressif/hal/src/esp32c3/bootloader_init.c (hunk 3 above)
Comment (almir-okato) [reply boot/espressif/hal/src/esp32c3/bootloader_init.c]: Fixed.

**Review by utzig — APPROVED** [review]
LGTM

**Pull request merged**